cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -2221,6 +2275,68 @@
 (rule (ud2 code)
       (SideEffectNoResult.Inst (MInst.Ud2 code)))
 
+;; Helper for creating `lzcnt` instructions.
+(decl lzcnt (Type Gpr) Gpr)
+(rule (lzcnt ty src)
+      (let ((dst WritableGpr (temp_writable_gpr))
+            (size OperandSize (operand_size_of_type_32_64 ty))
+            (_ Unit (emit (MInst.UnaryRmR size (UnaryRmROpcode.Lzcnt) src dst))))
+        dst))
+
+;; Helper for creating `tzcnt` instructions.
+(decl tzcnt (Type Gpr) Gpr)
+(rule (tzcnt ty src)
+      (let ((dst WritableGpr (temp_writable_gpr))
+            (size OperandSize (operand_size_of_type_32_64 ty))
+            (_ Unit (emit (MInst.UnaryRmR size (UnaryRmROpcode.Tzcnt) src dst))))
+        dst))
+
+;; Helper for creating `bsr` instructions.
+(decl bsr (Type Gpr) Gpr)
+(rule (bsr ty src)
+      (let ((dst WritableGpr (temp_writable_gpr))
+            (size OperandSize (operand_size_of_type_32_64 ty))
+            (_ Unit (emit (MInst.UnaryRmR size (UnaryRmROpcode.Bsr) src dst))))
+        dst))
+
+;; Helper for creating `bsr + cmov` instruction pairs that produce the
+;; result of the `bsr`, or `alt` if the input was zero.
+(decl bsr_or_else (Type Gpr Gpr) Gpr)
+(rule (bsr_or_else ty src alt)
+      (let ((dst WritableGpr (temp_writable_gpr))
+            (tmp WritableGpr (temp_writable_gpr))
+            (size OperandSize (operand_size_of_type_32_64 ty))
+            (_1 Unit (emit (MInst.UnaryRmR size (UnaryRmROpcode.Bsr) src tmp)))
+            (_2 Unit (emit (MInst.Cmove size (CC.Z) alt tmp dst))))
```

Comment:
The `CMOV` here uses `BSR`'s flag-setting behavior to work correctly. Just to check: we don't need to use `ConsumesFlags/ProducesFlags` here because these instructions are back-to-back and emitted in order? If that is the case, is a comment to that effect useful here?

Reply:
(Also, if this is allowed, then shouldn't the new `CmoveOr` and `XmmCmoveOr` sequences have been written in ISLE?)

Reply:
Hmm, yeah, I realize now that this is exactly the same circumstance that led me to suggest `CmoveOr` previously; the output of the bsr is one of the inputs to the cmove. I'm thinking now maybe we can add a new variant to `ConsumesFlags` for this; I'll play with it a bit, and clean up by removing `CmoveOr` at the same time if I can work something out. Sorry for the inconsistency and churn!

Reply:
OK, so I've reworked things so that `CmoveOr`/`XmmCmoveOr` no longer exist. Thanks for noticing this! The `i128` case needs a new `ConsumesFlags` variant that contains four consumer instructions, but this seems reasonable-ish to me.